### PR TITLE
[BUILD] DLL export interface for Metrics

### LINKED ISSUE #2313: DLL export interface for Metrics
### Discussed in https://github.com/open-telemetry/opentelemetry-cpp/discussions/2311

<div type='discussions-op-text'>

<sup>Originally posted by **TomSvk88** September 14, 2023</sup>
Hi,

I can see from the code itself that Traces and Logs classes as exported as dll interface but Metrics classes are not. Is there any particular reason for that or is it just it was not added yet?
Thank you for any response.

Tomas</div>

### PR DESCRIPTION
**Pull request by perhapsmaple**
Fixes #2313 

## Changes

Add Metrics classes to exported DLL interface.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

File: examples/common/metrics_foo_library/CMakeLists.txt
Comment (owent) [inline examples/common/metrics_foo_library/CMakeLists.txt:4]: This is not a problem of this PR, but I think we should use `if(OPENTELEMETRY_BUILD_DLL)` instead of `if(DEFINED OPENTELEMETRY_BUILD_DLL)`. And use `target_compile_definitions(opentelemetry_api INTERFACE OPENTELEMETRY_BUILD_IMPORT_DLL)` instead of `add_definitions(-DOPENTELEMETRY_BUILD_IMPORT_DLL)`, so we can exporting this definition to cmake package.

**Review by owent — APPROVED** [review]

File: ext/src/dll/CMakeLists.txt
Comment (ThomsonTan) [inline ext/src/dll/CMakeLists.txt:30]: Is this duplicated with line 14 above?

Comment (perhapsmaple) [reply ext/src/dll/CMakeLists.txt]: Yes, an oversight from my end. Fixed in the latest commit

**Comment by lalitb** [comment]
The current CI is only executing traces example as here - https://github.com/open-telemetry/opentelemetry-cpp/blob/e918960e06b45e899cd6e1f361acdb578d968334/ci/do_ci.ps1#L77

Should we also similarly add executing the metrics example, to ensure that the example is using with DLL.

**Comment by perhapsmaple** [comment]
Hi, I have added both the metrics and logs tests to the cmake.dll.test workflow. Let me know if any other changes are required. Thanks

**Review by lalitb — APPROVED** [review]
Thanks. Good to have @ThomsonTan approval before merging, as he wrote the initial DLL code.

**Review by ThomsonTan — APPROVED** [review]
Thanks for the contribution.

**Comment by meastp** [comment]
@perhapsmaple thanks a lot for working on this! we discovered this today while trying to get metrics to work - we are however having some build issues. Could you have a look, perhaps you can help us solve it? :) https://github.com/open-telemetry/opentelemetry-cpp/issues/2202#issuecomment-1782639249

**Pull request merged**
